language/stdlib/modules/Roles.move

```diff
@@ -0,0 +1,265 @@
+    public fun ASSOCIATION_ROOT_ROLE_ID(): u64 { 0 }
+    public fun TREASURY_COMPLIANCE_ROLE_ID(): u64 { 1 }
+    public fun DESIGNATED_DEALER_ROLE_ID(): u64 { 2 }
+    public fun VALIDATOR_ROLE_ID(): u64 { 3 }
+    public fun VALIDATOR_OPERATOR_ROLE_ID(): u64 { 4 }
+    public fun PARENT_VASP_ROLE_ID(): u64 { 5 }
+    public fun CHILD_VASP_ROLE_ID(): u64 { 6 }
+    public fun UNHOSTED_ROLE_ID(): u64 { 7 }
+
+    /// The roleId contains the role id for the account. This is only moved
+    /// to an account as a top-level resource, and is otherwise immovable.
+    /// INVARIANT: Once an account at address `A` is granted a role `R` it
+    ///            will remain an account with role `R` for all time.
+    resource struct RoleId {
+        role_id: u64,
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Privileges & Capabilities
+    ///////////////////////////////////////////////////////////////////////////
+
+    /// Privileges are extracted in to capabilities. Capabilities hold /
+    /// the account address that they were extracted from (i.e. tagged or
+    /// "tainted"). Capabilities can then only be restored to the account
+    /// where they were extracted from.
+    resource struct Capability<Privilege: resource> {
+        owner_address: address,
+    }
+
+    /// The internal representation of of a privilege. We wrap every
+    /// privilege witness resource here to avoid having to write extractors/restorers
+    /// for each privilege, but can instead write this generically.
+    resource struct Privilege<Priv: resource>  {
+        witness: Priv
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Role-specific Privileges
+    ///////////////////////////////////////////////////////////////////////////
+
+    /// Every role is granted a "default privilege" for that role. This can
+    /// be seen as a base-permission for every account of that role type.
+    /// INVARIANT: Every account has exactly one of these, and these
+    ///            correspond precisely to the RoleId.
+    resource struct AssociationRootRole {}
+    resource struct TreasuryComplianceRole {}
+    resource struct DesignatedDealerRole {}
+    resource struct ValidatorRole {}
+    resource struct ValidatorOperatorRole {}
+    resource struct ParentVASPRole {}
+    resource struct ChildVASPRole {}
+    resource struct UnhostedRole {}
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Privilege Granting
+    ///////////////////////////////////////////////////////////////////////////
+
+    /// The privilege `witness: Priv` is granted to `account` as long as
+    /// `account` has a `role` with `role.role_id == role_id`.
+    /// INVARIANT 1: Once a privilege witness `Priv` has been granted to an
+    ///              account it remains at that account.
+    public fun add_privilege_to_account<Priv: resource>(
```

Comment:
Wondering: can we make this private and expose wrappers without `role_id` for each role? That would also let us make the `ROLE_ID()` functions private (I think).

Reply:
ahhhh I was about to disagree, but I actually kinda like this idea a lot. Will change!